modules/config/main.tf

```diff
@@ -0,0 +1,304 @@
+locals {
+  cloudtrail_name = "${var.name}-config-trail"
+
+  cloudtrail_prefix      = var.cloudtrail_prefix == "" ? "gitlab-runner-config" : var.cloudtrail_prefix
+  cloudtrail_bucket      = length(var.cloudtrail_bucket) > 0 ? data.aws_s3_bucket.cloudtrail[0].bucket : "${var.name}-cloudtrail"
+  cloudtrail_bucket_name = length(var.cloudtrail_bucket) > 0 ? data.aws_s3_bucket.cloudtrail[0].bucket : aws_s3_bucket.cloudtrail[0].bucket
+  cloudtrail_bucket_arn  = "arn:aws:s3:::${local.cloudtrail_bucket}"
```

Comment:
Similarly to comment on `config_uri`, creating this name by hand means that we don't have a dependency visible to terraform. In case we leave `var.config_bucket` empty, we end up using `${var.name}-cloudtrail` as the value here and we end up with static string. Now terraform tries to make the `resource "aws_cloudtrail" "cloudtrail"` using this `local.config_bucket_arn` value as part of the policy. We _do_ create the bucket but it tends to take longer to do than making the cloudtrail policy. So AWS sees the ARN doesn't refer to a real bucket and we end up with:
```
Error: Error creating CloudTrail: InsufficientS3BucketPolicyException: Incorrect S3 bucket policy is detected for bucket: …-cloudtrail
```

What has to happen is we take this ARN from an actual terraform resource to enforce proper ordering.

So something like

```hcl
cloudtrail_bucket_arn  = length(var.cloudtrail_bucket) > 0 ? data.aws_s3_bucket.cloudtrail[0].arn : aws_s3_bucket.cloudtrail[0].arn
```

This then means that we have a cycle as the bucket policy mentions the ARN which we get from the bucket... so we have to resolve that via `aws_s3_bucket_policy` and attach the policy after bucket creation. This gets us back to the original problem: we create the cloudtrail before some resource is ready (before it was the bucket, now it's the policy). So we also need a `depends_on` on the `aws_s3_bucket_policy` in the cloudtrail.

With all these pieces in place, `terraform apply` just works again.

You can see all the changes at https://github.com/tsurucapital/terraform-aws-gitlab-runner/commit/751420121ade306cdebbb9d6eb186822579bf301